driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java

```diff
@@ -118,8 +118,9 @@ public void commitTransaction(final SingleResultCallback<Void> callback) {
             if (readConcern == null) {
                 throw new MongoInternalException("Invariant violated.  Transaction options read concern can not be null");
             }
+            boolean alreadyCommitted = commitInProgress || transactionState == TransactionState.COMMITTED;
```

Comment:
I'm not clear why `commitInProgress` in included in this condition.  What scenario does it cover?

Reply:
because it could be true and the state not yet set to COMMITTED.

Reply:
What I mean is, that flag was just added for #hasActiveTransaction method to work properly.  The only way that flag would be true here is if the app called this method concurrently from two different threads, which would be bad anyway. So for clarity I'd prefer to remove it and just check the transaction state.

Reply:
The race is if a user does:

```
session.commitTransaction(callBack);
session.commitTransaction(callBack);
```

Reply:
Regarding the race: if the user runs commitTransaction multiple times concurrently on the same transaction, they have bigger problems than the write concern. Results are undefined for concurrent use of the same transaction/session.

Regarding the state: it can happen when commitTransaction is retried, which is allowable according to spec.

Reply:
Yes, would potentially have issues - but I still think it is valid to check as per the spec for async.